File: rust/src/benchmarks/neural_net.rs
use super::super::{helper, Benchmark};

const INPUT_00: [f64; 2] = [0.0, 0.0];
const INPUT_01: [f64; 2] = [0.0, 1.0];
const INPUT_10: [f64; 2] = [1.0, 0.0];
const INPUT_11: [f64; 2] = [1.0, 1.0];
const TARGET_0: [f64; 1] = [0.0];
const TARGET_1: [f64; 1] = [1.0];

#[derive(Clone)]
struct Synapse {
    weight: f64,
    prev_weight: f64,

    source_neuron: usize,
    dest_neuron: usize,
}

impl Synapse {
    fn new(source_neuron: usize, dest_neuron: usize) -> Self {
        let w = helper::next_float(1.0) * 2.0 - 1.0;
        Self {
            weight: w,
            prev_weight: w,
            source_neuron,
            dest_neuron,
        }
    }
}

#[derive(Clone)]
struct Neuron {
    threshold: f64,
    prev_threshold: f64,
    output: f64,
    error: f64,
    synapses_in: Vec<usize>,
    synapses_out: Vec<usize>,
}

impl Neuron {
    const LEARNING_RATE: f64 = 1.0;
    const MOMENTUM: f64 = 0.3;

    fn new() -> Self {
        let t = helper::next_float(1.0) * 2.0 - 1.0;
        Self {
            threshold: t,
            prev_threshold: t,
            output: 0.0,
            error: 0.0,
            synapses_in: Vec::new(),
            synapses_out: Vec::new(),
        }
    }

    fn derivative(&self) -> f64 {
        self.output * (1.0 - self.output)
    }

    fn calculate_output(&mut self, sources: &[Neuron], synapses: &[Synapse]) {
        let mut activation = 0.0;
        for &synapse_idx in &self.synapses_in {
            let synapse = &synapses[synapse_idx];
            activation += synapse.weight * sources[synapse.source_neuron].output;
        }
        activation -= self.threshold;
        self.output = 1.0 / (1.0 + (-activation).exp());
    }

    fn update_weights(&mut self, rate: f64, sources: &[Neuron], synapses: &mut [Synapse]) {
        for &synapse_idx in &self.synapses_in {
            let synapse = &mut synapses[synapse_idx];
            let source_output = sources[synapse.source_neuron].output;
            let temp_weight = synapse.weight;
            synapse.weight += (rate * Self::LEARNING_RATE * self.error * source_output)
                + (Self::MOMENTUM * (synapse.weight - synapse.prev_weight));
            synapse.prev_weight = temp_weight;
        }

        let temp_threshold = self.threshold;
        self.threshold += (rate * Self::LEARNING_RATE * self.error * -1.0)
            + (Self::MOMENTUM * (self.threshold - self.prev_threshold));
        self.prev_threshold = temp_threshold;
    }
}

#[derive(Clone)]
struct NeuralNetwork {
    input_layer: Vec<Neuron>,
    hidden_layer: Vec<Neuron>,
    output_layer: Vec<Neuron>,
    synapses: Vec<Synapse>,
}

impl NeuralNetwork {
    fn new(inputs: usize, hidden: usize, outputs: usize) -> Self {
        let mut input_layer: Vec<_> = (0..inputs).map(|_| Neuron::new()).collect();
        let mut hidden_layer: Vec<_> = (0..hidden).map(|_| Neuron::new()).collect();
        let mut output_layer: Vec<_> = (0..outputs).map(|_| Neuron::new()).collect();
        let mut synapses = Vec::new();

        Self::connect(&mut input_layer, &mut hidden_layer, &mut synapses);
        Self::connect(&mut hidden_layer, &mut output_layer, &mut synapses);

        Self {
            input_layer,
            hidden_layer,
            output_layer,
            synapses,
        }
    }

    fn connect(sources: &mut [Neuron], destinations: &mut [Neuron], synapses: &mut Vec<Synapse>) {
        for (source_idx, source) in sources.iter_mut().enumerate() {
            for (dest_idx, dest) in destinations.iter_mut().enumerate() {
                let synapse_idx = synapses.len();
                synapses.push(Synapse::new(source_idx, dest_idx));
                source.synapses_out.push(synapse_idx);
                dest.synapses_in.push(synapse_idx);
            }
        }
    }

    fn train(&mut self, inputs: &[f64], targets: &[f64]) {
        self.feed_forward(inputs);

        const RATE: f64 = 0.3;

        for (neuron, &target) in self.output_layer.iter_mut().zip(targets) {
            neuron.error = (target - neuron.output) * neuron.derivative();
            neuron.update_weights(RATE, &self.hidden_layer, &mut self.synapses);
        }

        for neuron in &mut self.hidden_layer {
            let mut sum = 0.0;
            for &synapse_idx in &neuron.synapses_out {
                let synapse = &self.synapses[synapse_idx];
                sum += synapse.prev_weight * self.output_layer[synapse.dest_neuron].error;
            }
            neuron.error = sum * neuron.derivative();
            neuron.update_weights(RATE, &self.input_layer, &mut self.synapses);
        }
    }

    fn feed_forward(&mut self, inputs: &[f64]) {
        for (neuron, &input) in self.input_layer.iter_mut().zip(inputs) {
            neuron.output = input;
        }

        for neuron in &mut self.hidden_layer {
            neuron.calculate_output(&self.input_layer, &self.synapses);
        }

        for neuron in &mut self.output_layer {
            neuron.calculate_output(&self.hidden_layer, &self.synapses);
        }
    }

    fn current_outputs(&self) -> Vec<f64> {
        self.output_layer
            .iter()
            .map(|neuron| neuron.output)
            .collect()
    }
}

pub struct NeuralNet {
    xor_net: NeuralNetwork,
}

impl NeuralNet {
    pub fn new() -> Self {
        Self {
            xor_net: NeuralNetwork::new(0, 0, 0),
        }
    }
}

impl Benchmark for NeuralNet {
    fn name(&self) -> String {
        "Etc::NeuralNet".to_string()
    }

    fn prepare(&mut self) {
        self.xor_net = NeuralNetwork::new(2, 10, 1);
    }

    fn run(&mut self, _iteration_id: i64) {
        for _ in 0..1000 {
            self.xor_net.train(&INPUT_00, &TARGET_0);
            self.xor_net.train(&INPUT_10, &TARGET_1);
            self.xor_net.train(&INPUT_01, &TARGET_1);
            self.xor_net.train(&INPUT_11, &TARGET_0);
        }
    }

    fn checksum(&self) -> u32 {
        let mut net_copy = self.xor_net.clone();

        net_copy.feed_forward(&[0.0, 0.0]);
        let outputs1 = net_copy.current_outputs();

        net_copy.feed_forward(&[0.0, 1.0]);
        let outputs2 = net_copy.current_outputs();

        net_copy.feed_forward(&[1.0, 0.0]);
        let outputs3 = net_copy.current_outputs();

        net_copy.feed_forward(&[1.0, 1.0]);
        let outputs4 = net_copy.current_outputs();

        let mut all_outputs = Vec::new();
        all_outputs.extend_from_slice(&outputs1);
        all_outputs.extend_from_slice(&outputs2);
        all_outputs.extend_from_slice(&outputs3);
        all_outputs.extend_from_slice(&outputs4);

        let sum: f64 = all_outputs.iter().sum();
        helper::checksum_f64(sum)
    }
}

File: julia/src/benchmarks/NeuralNet.jl
mutable struct Synapse
    weight::Float64
    prev_weight::Float64
    source_neuron::Int32
    dest_neuron::Int32
end

function Synapse(source_neuron::Int32, dest_neuron::Int32)
    w = Helper.next_float() * 2.0 - 1.0
    return Synapse(w, w, source_neuron, dest_neuron)
end

mutable struct Neuron
    threshold::Float64
    prev_threshold::Float64
    output::Float64
    error::Float64
    synapses_in::Vector{Int32}
    synapses_out::Vector{Int32}
end

function Neuron()
    t = Helper.next_float() * 2.0 - 1.0
    return Neuron(t, t, 0.0, 0.0, Int32[], Int32[])
end

@inline function derivative(neuron::Neuron)::Float64
    return neuron.output * (1.0 - neuron.output)
end

const LEARNING_RATE = 1.0
const MOMENTUM = 0.3
const TRAIN_RATE = 0.3

mutable struct NeuralNetwork
    input_layer::Vector{Int32}
    hidden_layer::Vector{Int32}
    output_layer::Vector{Int32}
    neurons::Vector{Neuron}
    synapses::Vector{Synapse}
end

function NeuralNetwork(inputs::Int, hidden::Int, outputs::Int)
    total_neurons = inputs + hidden + outputs
    neurons = Vector{Neuron}(undef, total_neurons)

    for i = 1:total_neurons
        neurons[i] = Neuron()
    end

    input_layer = [i for i = 1:inputs]
    hidden_layer = [i for i = (inputs+1):(inputs+hidden)]
    output_layer = [i for i = (inputs+hidden+1):(inputs+hidden+outputs)]

    synapses = Synapse[]

    for source_idx in input_layer
        for dest_idx in hidden_layer
            synapse = Synapse(Int32(source_idx), Int32(dest_idx))
            push!(synapses, synapse)
            synapse_idx = length(synapses)

            push!(neurons[source_idx].synapses_out, Int32(synapse_idx))
            push!(neurons[dest_idx].synapses_in, Int32(synapse_idx))
        end
    end

    for source_idx in hidden_layer
        for dest_idx in output_layer
            synapse = Synapse(Int32(source_idx), Int32(dest_idx))
            push!(synapses, synapse)
            synapse_idx = length(synapses)

            push!(neurons[source_idx].synapses_out, Int32(synapse_idx))
            push!(neurons[dest_idx].synapses_in, Int32(synapse_idx))
        end
    end

    return NeuralNetwork(input_layer, hidden_layer, output_layer, neurons, synapses)
end

function train!(net::NeuralNetwork, inputs::Vector{Float64}, targets::Vector{Float64})
    feed_forward!(net, inputs)

    @inbounds for (i, target) in enumerate(targets)
        neuron_idx = net.output_layer[i]
        neuron = net.neurons[neuron_idx]

        error = (target - neuron.output) * derivative(neuron)
        net.neurons[neuron_idx].error = error

        for synapse_idx in neuron.synapses_in
            synapse = net.synapses[synapse_idx]
            source_output = net.neurons[synapse.source_neuron].output

            temp_weight = synapse.weight
            delta = TRAIN_RATE * LEARNING_RATE * error * source_output
            momentum_term = MOMENTUM * (synapse.weight - synapse.prev_weight)
            net.synapses[synapse_idx] = Synapse(
                synapse.weight + delta + momentum_term,
                temp_weight,
                synapse.source_neuron,
                synapse.dest_neuron,
            )
        end

        temp_threshold = neuron.threshold
        delta_threshold = TRAIN_RATE * LEARNING_RATE * error * -1.0
        momentum_threshold = MOMENTUM * (neuron.threshold - neuron.prev_threshold)
        net.neurons[neuron_idx] = Neuron(
            neuron.threshold + delta_threshold + momentum_threshold,
            temp_threshold,
            neuron.output,
            error,
            neuron.synapses_in,
            neuron.synapses_out,
        )
    end

    hidden_errors = Vector{Float64}(undef, length(net.hidden_layer))

    @inbounds for (i, neuron_idx) in enumerate(net.hidden_layer)
        neuron = net.neurons[neuron_idx]
        sum_error = 0.0

        for synapse_idx in neuron.synapses_out
            synapse = net.synapses[synapse_idx]

            sum_error += synapse.prev_weight * net.neurons[synapse.dest_neuron].error
        end

        hidden_errors[i] = sum_error * derivative(neuron)
    end

    @inbounds for (i, neuron_idx) in enumerate(net.hidden_layer)
        error = hidden_errors[i]
        neuron = net.neurons[neuron_idx]
        net.neurons[neuron_idx].error = error

        for synapse_idx in neuron.synapses_in
            synapse = net.synapses[synapse_idx]
            source_output = net.neurons[synapse.source_neuron].output

            temp_weight = synapse.weight
            delta = TRAIN_RATE * LEARNING_RATE * error * source_output
            momentum_term = MOMENTUM * (synapse.weight - synapse.prev_weight)
            net.synapses[synapse_idx] = Synapse(
                synapse.weight + delta + momentum_term,
                temp_weight,
                synapse.source_neuron,
                synapse.dest_neuron,
            )
        end

        temp_threshold = neuron.threshold
        delta_threshold = TRAIN_RATE * LEARNING_RATE * error * -1.0
        momentum_threshold = MOMENTUM * (neuron.threshold - neuron.prev_threshold)
        net.neurons[neuron_idx] = Neuron(
            neuron.threshold + delta_threshold + momentum_threshold,
            temp_threshold,
            neuron.output,
            error,
            neuron.synapses_in,
            neuron.synapses_out,
        )
    end
end

function feed_forward!(net::NeuralNetwork, inputs::Vector{Float64})

    @inbounds for (i, input) in enumerate(inputs)
        neuron_idx = net.input_layer[i]
        net.neurons[neuron_idx] = Neuron(
            net.neurons[neuron_idx].threshold,
            net.neurons[neuron_idx].prev_threshold,
            input,
            net.neurons[neuron_idx].error,
            net.neurons[neuron_idx].synapses_in,
            net.neurons[neuron_idx].synapses_out,
        )
    end

    @inbounds for neuron_idx in net.hidden_layer
        neuron = net.neurons[neuron_idx]
        activation = 0.0

        for synapse_idx in neuron.synapses_in
            synapse = net.synapses[synapse_idx]
            activation += synapse.weight * net.neurons[synapse.source_neuron].output
        end
        activation -= neuron.threshold

        output = 1.0 / (1.0 + exp(-activation))
        net.neurons[neuron_idx] = Neuron(
            neuron.threshold,
            neuron.prev_threshold,
            output,
            neuron.error,
            neuron.synapses_in,
            neuron.synapses_out,
        )
    end

    @inbounds for neuron_idx in net.output_layer
        neuron = net.neurons[neuron_idx]
        activation = 0.0

        for synapse_idx in neuron.synapses_in
            synapse = net.synapses[synapse_idx]
            activation += synapse.weight * net.neurons[synapse.source_neuron].output
        end
        activation -= neuron.threshold

        output = 1.0 / (1.0 + exp(-activation))
        net.neurons[neuron_idx] = Neuron(
            neuron.threshold,
            neuron.prev_threshold,
            output,
            neuron.error,
            neuron.synapses_in,
            neuron.synapses_out,
        )
    end
end

function current_outputs(net::NeuralNetwork)::Vector{Float64}
    outputs = Vector{Float64}(undef, length(net.output_layer))
    @inbounds for (i, neuron_idx) in enumerate(net.output_layer)
        outputs[i] = net.neurons[neuron_idx].output
    end
    return outputs
end

mutable struct NeuralNet <: AbstractBenchmark
    xor_net::Union{NeuralNetwork,Nothing}
    result::UInt32

    function NeuralNet()
        new(nothing, UInt32(0))
    end
end

name(b::NeuralNet)::String = "Etc::NeuralNet"

function prepare(b::NeuralNet)

    b.xor_net = NeuralNetwork(2, 10, 1)
end

function run(b::NeuralNet, iteration_id::Int64)
    if b.xor_net === nothing
        error("Neural network not initialized. Call prepare() first.")
    end

    train!(b.xor_net, [0.0, 0.0], [0.0])
    train!(b.xor_net, [1.0, 0.0], [1.0])
    train!(b.xor_net, [0.0, 1.0], [1.0])
    train!(b.xor_net, [1.0, 1.0], [0.0])
end

function checksum(b::NeuralNet)::UInt32
    if b.xor_net === nothing
        return UInt32(0)
    end

    net = b.xor_net

    total = 0.0

    feed_forward!(net, [0.0, 0.0])
    total += Base.sum(current_outputs(net))

    feed_forward!(net, [0.0, 1.0])
    total += Base.sum(current_outputs(net))

    feed_forward!(net, [1.0, 0.0])
    total += Base.sum(current_outputs(net))

    feed_forward!(net, [1.0, 1.0])
    total += Base.sum(current_outputs(net))

    return Helper.checksum_f64(total)
end
